source/adapters/level_zero/v2/event_pool_cache.hpp

```diff
@@ -0,0 +1,47 @@
+//===--------- event_pool_cache.hpp - Level Zero Adapter ------------------===//
+//
+// Copyright (C) 2024 Intel Corporation
+//
+// Part of the Unified-Runtime Project, under the Apache License v2.0 with LLVM
+// Exceptions. See LICENSE.TXT
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+#pragma once
+
+#include <functional>
+#include <memory>
+#include <mutex>
+#include <stack>
+
+#include <unordered_map>
+#include <ur/ur.hpp>
+#include <ur_api.h>
+#include <ze_api.h>
+
+#include "../device.hpp"
+#include "event_pool.hpp"
+#include "event_provider.hpp"
+
+namespace v2 {
+
+using event_pool_borrowed =
+    std::unique_ptr<event_pool, std::function<void(event_pool *)>>;
+
+class event_pool_cache {
+public:
+  using ProviderCreateFunc =
+      std::function<std::unique_ptr<event_provider>(DeviceId)>;
+
+  event_pool_cache(ProviderCreateFunc);
+  ~event_pool_cache();
+
+  event_pool_borrowed borrow(DeviceId);
+
+private:
+  ur_mutex Mutex;
+  ProviderCreateFunc ProviderCreate;
+  std::unordered_map<DeviceId, std::vector<std::unique_ptr<event_pool>>> Pools;
```

Comment:
I don;t think you need to use unique_ptr here, you should be able to just store event_pool directly.

Reply:
Also, do we actually need unordered_map here? Vector should work too I think

Reply:
> I don;t think you need to use unique_ptr here, you should be able to just store event_pool directly.

I sort this vector and wanted to avoid moving possibly large structs.

> Also, do we actually need unordered_map here? Vector should work too I think

Done, but this will require us to specify the max number of devices when constructing the cache.